velox/substrait/VeloxToSubstraitExpr.cpp

```diff
@@ -0,0 +1,452 @@
+
+    return *substraitExpr;
+  } else {
+    VELOX_UNSUPPORTED("Unsupport Expr '{}' in Substrait", expr->toString());
+  }
+}
+
+const ::substrait::Expression_Cast&
+VeloxToSubstraitExprConvertor::toSubstraitExpr(
+    google::protobuf::Arena& arena,
+    const std::shared_ptr<const CastTypedExpr>& castExpr,
+    const RowTypePtr& inputType) {
+  ::substrait::Expression_Cast* substraitCastExpr =
+      google::protobuf::Arena::CreateMessage<::substrait::Expression_Cast>(
+          &arena);
+  std::vector<std::shared_ptr<const ITypedExpr>> castExprInputs =
+      castExpr->inputs();
+
+  substraitCastExpr->mutable_type()->MergeFrom(
+      typeConvertor_.toSubstraitType(arena, castExpr->type()));
+
+  for (auto& arg : castExprInputs) {
+    substraitCastExpr->mutable_input()->MergeFrom(
+        toSubstraitExpr(arena, arg, inputType));
+  }
+  return *substraitCastExpr;
+}
+
+const ::substrait::Expression_FieldReference&
+VeloxToSubstraitExprConvertor::toSubstraitExpr(
+    google::protobuf::Arena& arena,
+    const std::shared_ptr<const FieldAccessTypedExpr>& fieldExpr,
+    const RowTypePtr& inputType) {
+  ::substrait::Expression_FieldReference* substraitFieldExpr =
+      google::protobuf::Arena::CreateMessage<
+          ::substrait::Expression_FieldReference>(&arena);
+
+  const auto& exprType = fieldExpr->type();
+  std::string exprName = fieldExpr->name();
+
+  ::substrait::Expression_ReferenceSegment_StructField* directStruct =
+      substraitFieldExpr->mutable_direct_reference()->mutable_struct_field();
+
+  const auto& preNodeColNames = inputType->names();
+  const auto& preNodeColTypes = inputType->children();
+  const auto& preNodeColNums = preNodeColNames.size();
+  auto currentColId = -1;
+
+  VELOX_CHECK_EQ(preNodeColNums, preNodeColTypes.size());
+
+  for (auto i = 0; i < preNodeColNums; i++) {
+    if (preNodeColNames[i] == exprName && preNodeColTypes[i] == exprType) {
+      currentColId = i;
+      break;
+    }
+  }
+
+  if (currentColId == -1) {
+    currentColId = preNodeColNums + 1;
+  }
+  directStruct->set_field(currentColId);
+
+  return *substraitFieldExpr;
+}
+
+const ::substrait::Expression& VeloxToSubstraitExprConvertor::toSubstraitExpr(
+    google::protobuf::Arena& arena,
+    const std::shared_ptr<const CallTypedExpr>& callTypeExpr,
+    const RowTypePtr& inputType) {
+  ::substrait::Expression* substraitExpr =
+      google::protobuf::Arena::CreateMessage<::substrait::Expression>(&arena);
+
+  auto callTypeExprInputs = callTypeExpr->inputs();
+  std::string callTypeExprFunName = callTypeExpr->name();
+
+  // The processing is different for different function names.
+  // TODO add support for if Expr and switch Expr
+  if (callTypeExprFunName != exec::kIf &&
+      callTypeExprFunName != exec::kSwitch) {
+    ::substrait::Expression_ScalarFunction* scalaExpr =
+        substraitExpr->mutable_scalar_function();
+
+    // TODO need to change yaml file to register function, now is dummy.
```

Comment:
registerSubstraitFunction increments a counter and returns current value. If called multiple times with the same name it will return different values. Is this desired?

Instead, shouldn't we add some setup logic to assign each registered Velox function an ID and then look up that ID here?

Reply:
> Instead, shouldn't we add some setup logic to assign each registered Velox function an ID and then look up that ID here?

Yes, we will split register and lookup later.